Mathlib/Init/Control/Lawful.lean

```diff
@@ -0,0 +1,246 @@
+/-
+Copyright (c) 2017 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Ullrich
+
+! This file was ported from Lean 3 source module init.control.lawful
+! leanprover-community/lean commit 9af482290ef68e8aaa5ead01aa7b09b7be7019fd
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+
+import Mathlib.Mathport.Rename
+import Mathlib.Tactic.Basic
+
+/-! ## Functor Laws, applicative laws, and monad Laws -/
+
+universe u v
+
+#align is_lawful_functor LawfulFunctor
+#align is_lawful_functor.map_const_eq LawfulFunctor.map_const
+#align is_lawful_functor.id_map LawfulFunctor.id_map
+#align is_lawful_functor.comp_map LawfulFunctor.comp_map
+
+#align is_lawful_applicative LawfulApplicative
+#align is_lawful_applicative.seq_left_eq LawfulApplicative.seqLeft_eq
+#align is_lawful_applicative.seq_right_eq LawfulApplicative.seqRight_eq
+#align is_lawful_applicative.pure_seq_eq_map LawfulApplicative.pure_seq
+#align is_lawful_applicative.map_pure LawfulApplicative.map_pure
+#align is_lawful_applicative.seq_pure LawfulApplicative.seq_pure
+#align is_lawful_applicative.seq_assoc LawfulApplicative.seq_assoc
+
+#align pure_id_seq pure_id_seq
+
+#align is_lawful_monad LawfulMonad
+#align is_lawful_monad.bind_pure_comp_eq_map LawfulMonad.bind_pure_comp
+#align is_lawful_monad.bind_map_eq_seq LawfulMonad.bind_map
+#align is_lawful_monad.pure_bind LawfulMonad.pure_bind
+#align is_lawful_monad.bind_assoc LawfulMonad.bind_assoc
+
+#align bind_pure bind_pure
+
+#align bind_ext_congr bind_congr
+
+#align map_ext_congr map_congr
+
+#align id.map_eq Id.map_eq
+
+#align id.bind_eq Id.bind_eq
+
+#align id.pure_eq Id.pure_eq
+
+namespace StateT
+
+section
+
+variable {σ : Type u}
+
+variable {m : Type u → Type v}
+
+variable {α : Type u}
+
+/-
+Porting note:
+In Lean 4, `StateT` doesn't require a constructor, but it appears confusing to declare the
+following theorem as a simp theorem.
```

Comment:
Sorry, what is wrong with this?

Reply:
If we decleare this theorem as a simp theorem, `StateT.run f st` is simplified to `f st` by eta reduction. This breaks the structure of `StateT`. `ReaderT` also.
I've added the porting note.